### PR TITLE
Pause to disable dependency gathering

### LINKED ISSUE #5072: Scheduler file doesn't like paths with ~
There are a few issues related to using `~` in the scheduler file option.

1. When starting a scheduler with a scheduler file you get a `FileNotFoundError`.

```
$ dask-scheduler --scheduler-file "~/scheduler.json"
distributed.scheduler - INFO - -----------------------------------------------
distributed.scheduler - INFO - -----------------------------------------------
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:   tcp://10.51.100.15:8786
distributed.scheduler - INFO -   dashboard at:                     :8787
distributed.scheduler - INFO - End scheduler at 'tcp://10.51.100.15:8786'
Traceback (most recent call last):
  File "/Users/jtomlinson/miniconda3/envs/dask/bin/dask-scheduler", line 33, in <module>
    sys.exit(load_entry_point('distributed', 'console_scripts', 'dask-scheduler')())
  File "/Users/jtomlinson/Projects/dask/distributed/distributed/cli/dask_scheduler.py", line 217, in go
    main()
  File "/Users/jtomlinson/miniconda3/envs/dask/lib/python3.8/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/Users/jtomlinson/miniconda3/envs/dask/lib/python3.8/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/Users/jtomlinson/miniconda3/envs/dask/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/jtomlinson/miniconda3/envs/dask/lib/python3.8/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/Users/jtomlinson/Projects/dask/distributed/distributed/cli/dask_scheduler.py", line 208, in main
    loop.run_sync(run)
  File "/Users/jtomlinson/miniconda3/envs/dask/lib/python3.8/site-packages/tornado/ioloop.py", line 530, in run_sync
    return future_cell[0].result()
  File "/Users/jtomlinson/Projects/dask/distributed/distributed/cli/dask_scheduler.py", line 204, in run
    await scheduler
  File "/Users/jtomlinson/Projects/dask/distributed/distributed/core.py", line 283, in _
    await self.start()
  File "/Users/jtomlinson/Projects/dask/distributed/distributed/scheduler.py", line 3791, in start
    with open(self.scheduler_file, "w") as f:
FileNotFoundError: [Errno 2] No such file or directory: '~/scheduler.json'
```

2. When connecting a worker or client to a scheduler with a `~` in the scheduler file path it hangs indefinitely.

```
$ dask-worker --scheduler-file "~/scheduler.json"

Hung...
```

3. A user has also reported issues with client but I have been unable to reproduce locally.  dask/dask-mpi#71

```python
In [1]: from dask.distributed import Client
cl
In [2]: client = Client(scheduler_file="~/scheduler.json")

In [3]: client
Out[3]: <Client: 'tcp://10.51.100.15:8786' processes=0 threads=0, memory=0 B>
```

### LINKED ISSUE #5702: Paused workers should not be able to fetch any more data
The current implementation of the worker `ensure_communicating` will continue to fetch dependencies for as long as there are dependencies to fetch. This can push an already overloaded worker over the edge and cause it to fail.

https://github.com/dask/distributed/blob/8734c9db78a68f6bea5d2a1deb39412a9e7b739a/distributed/worker.py#L2684-L2687

A paused worker should not be allowed to fetch more data.

There are two possible ways to achieve this

1. Add another guard to `ensure_communicating` to stop scheduling additional `gather_dep` coroutines
2. Remove all tasks from a paused worker that aren't in memory. This would indirectly empty the `data_needed` heap and cause a worker to stabilize. This could be achieved by either aggressively stealing or by implementing a custom scheduler handler.

I think both options have a certain appeal. I'm wondering which one is the best to choose, specifically in context of the latest changes to AMM / retirement / pause. 

cc @crusaderky 

Note: right now, network traffic is only restricted for egress, i.e. incoming `get_data` requests from other workers, see https://github.com/dask/distributed/blob/8734c9db78a68f6bea5d2a1deb39412a9e7b739a/distributed/worker.py#L1712-L1716

### PR DESCRIPTION
**Pull request by crusaderky**
Closes #5702

### CODE REVIEW

File: distributed/tests/test_worker_memory.py [hunk 1]
```diff
@@ -478,6 +478,71 @@ def f(ev):
...
+                if w.address == a_addr:
+                    w.status = Status.paused
+                    return X()
+            assert False
+
+    x = c.submit(X, key="x", workers=[b.address])
+    y = c.submit(inc, 1, key="y", workers=[b.address])
+    await wait([x, y])
+
+    # - z reaches worker a with higher priority than w
```
Comment (crusaderky) [inline distributed/tests/test_worker_memory.py:516]: As far as I could test, this seems to be deterministic - however I can't understand if it's intentional?

Comment (fjetter) [reply distributed/tests/test_worker_memory.py:516]: If you rely on the priorities in this test, I would suggest to set them explicitly. This not only makes the test easier to read but it is also more robust to changes in the way we calculate internal priorities, i.e. excluding user priorities.

File: distributed/tests/test_worker_memory.py [hunk 2]
```diff
@@ -478,6 +478,71 @@ def f(ev):
...
+    # - x flight->memory; a is removed from in_flight_workers
+    # - ensure_communicating is triggered again
+    # - ensure_communicating refuses to fetch y because the worker is paused
+
+    w = c.submit(inc, y, key="w", workers=[a.address])
+    z = c.submit(lambda _: None, x, key="z", workers=[a.address])
+
+    while "y" not in a.tasks or a.tasks["y"].state != "fetch":
+        await asyncio.sleep(0.01)
+    await asyncio.sleep(0.1)
```
Comment (fjetter) [inline distributed/tests/test_worker_memory.py:536]: why is this additional sleep required?

**Review by fjetter — APPROVED** [review]
I would suggest to use explicit priorities in the test. If priorities are not properly assigned on task submission that is obviously something we care about but this test should not care about that fact.

Otherwise, LGTM

File: distributed/tests/test_worker_memory.py (hunk 2 above)
Comment (crusaderky) [reply distributed/tests/test_worker_memory.py:536]: To make sure the task remained in fetch state and didn't transition to flight

**Pull request merged**